pkg/insights/insightsreport/insightsreport.go

```diff
@@ -0,0 +1,184 @@
+// Gatherer gathers the report from Smart Proxy
+type Gatherer struct {
+	controllerstatus.Simple
+
+	configurator          Configurator
+	client                *insightsclient.Client
+	LastReport            SmartProxyReport
+	archiveUploadReporter <-chan struct{}
+	retries               int64
+}
+
+// Response represents the Smart Proxy report response structure
+type Response struct {
+	Report SmartProxyReport `json:"report"`
+}
+
+// Configurator represents the interface to retrieve the configuration for the gatherer
+type Configurator interface {
+	Config() *config.Controller
+	ConfigChanged() (<-chan struct{}, func())
+}
+
+// InsightsReporter represents an object that can notify about archive uploading
+type InsightsReporter interface {
+	ArchiveUploaded() <-chan struct{}
+}
+
+var (
+	// ErrWaitingForVersion contains an error when the cluster version is not loaded yet
+	ErrWaitingForVersion = fmt.Errorf("waiting for the cluster version to be loaded")
+
+	// insightsStatus contains a metric with the latest report information
+	insightsStatus = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Namespace: "health",
+		Subsystem: "statuses",
+		Name:      "insights",
+		Help:      "Information about the cluster health status as detected by Insights tooling.",
+	}, []string{"metric"})
+)
+
+// New initializes and returns a Gatherer
+func New(client *insightsclient.Client, configurator Configurator, reporter InsightsReporter) *Gatherer {
+	return &Gatherer{
+		Simple:                controllerstatus.Simple{Name: "insightsreport"},
+		configurator:          configurator,
+		client:                client,
+		archiveUploadReporter: reporter.ArchiveUploaded(),
+		retries:               0,
+	}
+}
+
+// PullSmartProxy performs a request to the Smart Proxy and unmarshal the response
+func (r *Gatherer) PullSmartProxy() (bool, error) {
+	klog.Info("Pulling report from smart-proxy")
+	config := r.configurator.Config()
+	initialWait := config.ReportPullingDelay
+	minimalRetryTime := config.ReportMinRetryTime
+	reportEndpoint := config.ReportEndpoint
+
+	if len(reportEndpoint) == 0 {
+		klog.V(4).Info("Not downloading report because Smart Proxy client is not properly configured: missing report endpoint")
+		return true, nil
+	}
+
+	delay := initialWait - time.Duration(r.retries*minimalRetryTime.Nanoseconds()) - minimalRetryTime
+	if delay > 0 {
+		time.Sleep(delay)
+	}
+	r.retries++
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Minute)
+	defer cancelFunc()
+
+	klog.V(4).Info("Retrieving report")
+	reportBody, err := r.client.RecvReport(ctx, reportEndpoint)
+	if authorizer.IsAuthorizationError(err) {
+		r.Simple.UpdateStatus(controllerstatus.Summary{
+			Operation: controllerstatus.DownloadingReport,
+			Reason:    "NotAuthorized",
+			Message:   fmt.Sprintf("Auth rejected for downloading latest report: %v", err),
+		})
+		return false, err
+	} else if err != nil {
+		klog.Errorf("Error retrieving the report: %s", err)
+		return false, err
```

Comment:
Any other error should also stop retrying immediatelly.

Reply:
Not sure about this. I'll review the insightsclient code and check what errors can be returned

Reply:
If any other error then unauthorized happens,, it needs to return true, err to indicate that this function shouldn't be called again.